# HG changeset patch
# User Augie Fackler <raf@durin42.com>
# Date 1418753230 18000
#      Tue Dec 16 13:07:10 2014 -0500
# Branch stable
# Node ID 7a5bcd471f2ef302613b8551a79081d46d04be6e
# Parent  885bd7c5c7e3efc10081c09c11e538a3fa19ace4
darwin: omit ignorable codepoints when normcase()ing a file path

This lets us avoid some nasty case collision problems in OS X with
invisible codepoints.

--- a/mercurial/posix.py
+++ b/mercurial/posix.py
@@ -191,7 +191,9 @@ if sys.platform == 'darwin':
             u = s.decode('utf-8')
 
         # Decompose then lowercase (HFS+ technote specifies lower)
-        return unicodedata.normalize('NFD', u).lower().encode('utf-8')
+        enc = unicodedata.normalize('NFD', u).lower().encode('utf-8')
+        # drop HFS+ ignored characters
+        return encoding.hfsignoreclean(enc)
 
     def realpath(path):
         '''
--- a/tests/test-casefolding.t
+++ b/tests/test-casefolding.t
@@ -165,12 +165,11 @@ case changes.
 We assume anyone running the tests on a case-insensitive volume on OS
 X will be using HFS+. If that's not true, this test will fail.
 
-Bug: some codepoints are to be ignored on HFS+:
-
   $ rm A
   >>> open(u'a\u200c'.encode('utf-8'), 'w').write('unicode is fun')
   $ hg status
   M A
-  ? a\xe2\x80\x8c (esc)
+
 #endif
+
   $ cd ..
